contrib/babelfishpg_tsql/src/catalog.c

```diff
@@ -2648,6 +2653,102 @@ rename_update_bbf_catalog(RenameStmt *stmt)
 	}
 }
 
+static void
+rename_object_update_bbf_schema_permission_catalog(RenameStmt *stmt, int rename_type)
+{
+	/* Update 'object_name' in 'babelfish_schema_permissions' */
+	Relation	bbf_schema_rel;
+	TupleDesc	bbf_schema_dsc;
+	ScanKeyData key[4];
+	HeapTuple	tuple_bbf_schema;
+	HeapTuple	new_tuple;
+	SysScanDesc scan;
+	Datum		new_record_bbf_schema[BBF_SCHEMA_PERMS_NUM_OF_COLS] = {0};
+	bool		new_record_nulls_bbf_schema[BBF_SCHEMA_PERMS_NUM_OF_COLS] = {false};
+	bool		new_record_repl_bbf_schema[BBF_SCHEMA_PERMS_NUM_OF_COLS] = {false};
+	const char *logical_schema_name;
+	const char *object_name;
+	char	*object_type = NULL;
+	int16		dbid = get_cur_db_id();
+	Node	   *schema;
+	ObjectWithArgs *objwargs;
+
+	/* open the catalog table */
+	bbf_schema_rel = table_open(get_bbf_schema_perms_oid(), RowExclusiveLock);
+	/* get the description of the table */
+	bbf_schema_dsc = RelationGetDescr(bbf_schema_rel);
+
+	if (rename_type == OBJECT_TABLE || rename_type == OBJECT_VIEW)
+	{
+		logical_schema_name = get_logical_schema_name(stmt->relation->schemaname, true);
+		object_name = stmt->relation->relname;
+		object_type = OBJ_RELATION;
+	}
+	else
+	{
+		if (rename_type == OBJECT_PROCEDURE)
+			object_type = OBJ_PROCEDURE;
+		else if (rename_type == OBJECT_FUNCTION)
+			object_type = OBJ_FUNCTION;
+		objwargs = (ObjectWithArgs *) stmt->object;
+		schema = (Node *) linitial(objwargs->objname);
+		logical_schema_name = strVal(schema);
+		object_name = stmt->subname;
+	}
+
+	/* search for the row for update => build the key */
+	ScanKeyInit(&key[0],
+				Anum_bbf_schema_perms_dbid,
+				BTEqualStrategyNumber, F_INT2EQ,
+				Int16GetDatum(dbid));
+	ScanKeyEntryInitialize(&key[1], 0,
+				Anum_bbf_schema_perms_schema_name,
+				BTEqualStrategyNumber, InvalidOid,
+				tsql_get_server_collation_oid_internal(false),
+				F_TEXTEQ, CStringGetTextDatum(logical_schema_name));
+	ScanKeyEntryInitialize(&key[2], 0,
+				Anum_bbf_schema_perms_object_name,
+				BTEqualStrategyNumber, InvalidOid,
+				tsql_get_server_collation_oid_internal(false),
+				F_TEXTEQ, CStringGetTextDatum(object_name));
+	ScanKeyEntryInitialize(&key[3], 0,
+				Anum_bbf_schema_perms_object_type,
+				BTEqualStrategyNumber,
+				InvalidOid,
+				tsql_get_server_collation_oid_internal(false),
+				F_TEXTEQ,
+				CStringGetTextDatum(object_type));
+
+	/* scan */
+	scan = systable_beginscan(bbf_schema_rel,
+			get_bbf_schema_perms_idx_oid(),
+			true, NULL, 4, key);
+
+	/* get the scan result -> original tuple */
+	tuple_bbf_schema = systable_getnext(scan);
+
+	while (HeapTupleIsValid(tuple_bbf_schema))
```

Comment:
1. Can there be more than one entry? If not, better to use if condition for readability.
2. Let's add a comment why it's okay not to throw an error if an entry is not found.

Reply:
1. While is needed since there can be multiple rows.
2. Added.